File: CoreJava/SOLID-princples-app/src/srp/refractor/Invoice.java
package srp.refractor;

import java.util.*;

public class Invoice {
	private final String id;
	private final String name;
	private final double cost;
	private final double discount;
	private final static double GST = 5 ;

	public Invoice(String name, double cost, double discount) {
		this.id = UUID.randomUUID().toString().replace("-","");
		this.name = name;
		this.cost = cost;
		this.discount = discount;
	}

	public String getId() {
		return id;
	}

	public String getName() {
		return name;
	}

	public double getCost() {
		return cost;
	}

	public double getDiscount() {
		return discount;
	}

	public double getGST() {
		return GST;
	}

	protected double calcDiscount() {
		return Math.abs(this.getDiscount() - this.getCost());
	}

	protected double calcTax() {
		return (getGST() * getCost() - calcDiscount()) /100;
	}

	protected double calcTotal() {
		return (this.getCost() - this.getDiscount()) + this.calcTax();
	}	
}

File: DotNET/Solid Principles/SRPRefractorApp/SRPRefractorApp/Invoice.cs
using System;

namespace SRPRefractorApp
{
    class Invoice
    {
        private String _id;
        private String _name;
        private double _cost;
        private double _discount;
        private static double GST = 5;

        public Invoice(String id, String name, double cost, double discount)
        {
            _id = id;
            _name = name;
            _cost = cost;
            _discount = discount;
        }

        public String Id
        {
            get
            {
                return _id;
            }
        }

        public String Name
        {
            get
            {
                return _name;
            }
        }

        public double Cost
        {
            get
            {
                return _cost;
            }
        }

        public double Discount
        {
            get
            {
                return _discount;
            }
        }

        public double CalculateDiscount()
        {
            return Math.Abs(Discount - Cost);
        }

        public double CalculateTax()
        {
            return GST * (Cost - CalculateDiscount());
        }

        public double CalculateTotal()
        {
            return (Cost - Discount) + CalculateTax();
        }
    }
}
